### PR TITLE
hoist regexp literals in SourceMapDerToolPlugin

### LINKED ISSUE #15720: Hoist RegExp Literals in SourceMapDevToolPlugin
# Bug report

**What is the current behavior?**
The use of RegExp literals create unnecessary object references/allocations especially in iterated hot paths like SourceMapDevToolPlugin. We should hoist these to the top of file and give them useful names so people understand what the heck they do. 

**If the current behavior is a bug, please provide the steps to reproduce.**
Some Examples: 

Mapping over a RegExp literal: https://github.com/webpack/webpack/blob/3ad4fcac25a976277f2d9cceb37bc81602e96b13/lib/SourceMapDevToolPlugin.js#L419

Big Async ForEach Statement with multiple RexExp Literal statements: 
https://github.com/webpack/webpack/blob/3ad4fcac25a976277f2d9cceb37bc81602e96b13/lib/SourceMapDevToolPlugin.js#L384-L551

**What is the expected behavior?**
Less memory allocation, less garbage collection triggered.

Happy to PR this. There are quite a few of these examples still lying around webpack that should be cleaned up but I'll start with this.

### PR DESCRIPTION
**Pull request by TheLarkInn**
**What kind of change does this PR introduce?**
Hoists Regular Expression Literals into variables and gives them descriptive (?) names. 

Refactoring
**Did you add tests for your changes?**

**Does this PR introduce a breaking change?**
No

**What needs to be documented once your changes are merged?**
Nothing documented, however I would like a review on the names of the variables themselves so they can be as descriptive as possible.

Fixes #15720

### CODE REVIEW

**Comment by TheLarkInn** [comment]
I'll need a day or so to determine the right type of CLA signing.

**Comment by TheLarkInn** [comment]
@sokra can you spot-check these RegExp names? They probably need to be reworded.

File: lib/SourceMapDevToolPlugin.js [hunk 1]
```diff
@@ -47,13 +47,21 @@ const validate = createSchemaValidation(
  * @property {ItemCacheFacade} cacheItem cache item
  */
 
+const METACHARACTERS_REGEXP = /[-[\]\\/{}()*+?.^$|]/g;
```
Comment (dmichon-msft) [inline lib/SourceMapDevToolPlugin.js:50]: The one hazard of hoisting regexps is that they maintain internal state, so in some cases you'll need to set `.lastIndex = -1` before each time you use the regexp.

Comment (TheLarkInn) [reply lib/SourceMapDevToolPlugin.js:50]: Would you mind checking my changes here? Looks like `.test()` was only called in two places that would warrant this gaurd.

**Review by alexander-akait — APPROVED** [review]

**Comment by TheLarkInn** [comment]
We can force merge this we only have flaky coverage remaining.

**Pull request merged**